base/server/python/pki/ldapdb/__init__.py

```diff
@@ -0,0 +1,205 @@
+        provided earlier
+
+        :return: None
+        """
+
+        self.temp_dir = tempfile.mkdtemp()
+
+        if self.nssdb_dir:
+            ldap.set_option(ldap.OPT_X_TLS_CACERTDIR, self.nssdb_dir)
+
+        if self.client_cert_nickname:
+            password_file = os.path.join(self.temp_dir, 'password.txt')
+            with open(password_file, 'w') as f:
+                f.write(self.nssdb_password)
+
+            ldap.set_option(ldap.OPT_X_TLS_CERTFILE, self.client_cert_nickname)
+            ldap.set_option(ldap.OPT_X_TLS_KEYFILE, password_file)
+
+        self.ldap = ldap.initialize(self.url)
+
+        if self.bind_dn and self.bind_password:
+            self.ldap.simple_bind_s(self.bind_dn, self.bind_password)
+
+    def close(self):
+
+        if self.ldap:
+            self.ldap.unbind_s()
+
+        if self.temp_dir:
+            shutil.rmtree(self.temp_dir)
+
+    def get(self, base_dn, search_scope, search_filter, retrieve_attributes):
+        """
+        Retrieve entries from the LDAP database synchronously
+
+        :param base_dn:
+            The base Distinguished Name (DN) to start search
+        :type base_dn: str
+
+        :param search_scope:
+            The scope of this search (SCOPE_BASE, SCOPE_ONELEVEL, SCOPE_SUBTREE)
+        :type search_scope: int
+
+        :param search_filter:
+            The search filter string
+        :type search_filter: str
+
+        :param retrieve_attributes:
+            List of attributes to retrieve
+        :type retrieve_attributes: list
+
+        :return:
+            Result of the query
+        :rtype: list
+
+        """
+
+        # If ldap object isn't initialized, no operation can be performed
+        if not self.ldap:
+            raise Exception('LDAP instance uninitialized.')
+
+        try:
+            # Perform a synchronous search
+            ldap_result = self.ldap.search_s(
+                base_dn, search_scope, search_filter, retrieve_attributes)
+            return ldap_result
+
+        except ldap.LDAPError as e:
+            logger.error(e)
+
+    def put(self, dn, attr):
+        """
+        Add an entry to the LDAP database synchronously
+
+        :param dn:
+            The base Distinguished Name (DN) to start search
+        :type dn: str
+
+        :param attr:
+            The scope of this search (SCOPE_BASE, SCOPE_ONELEVEL, SCOPE_SUBTREE)
+        :type attr: dict
+
+        :return:
+            Result of the insert operation
+        :rtype: bool
+
+        """
+
+        if not self.ldap:
+            raise Exception('LDAP instance uninitialized.')
+
+        try:
+            logger.debug('Converting dict to LDAP entry syntax')
+            # Convert our dict to nice syntax for the add-function
+            ldif = modlist.addModlist(attr)
+
+            logger.debug('Adding entry to LDAP db')
+            # Perform a synchronous add operation
+            ldap_result = self.ldap.add_s(dn, ldif)
+
+            if ldap_result:
```

Comment:
According to the doc the add_s() method will always return None.

Reply:
Yes, I read that. But, it seems like add_s returns a tuple. Not sure whether it's a leak/bug in their module.

Reply:
Regardless, we should not rely on ldap.add_s() return value. If there's an error during LDAP Add operation the method will throw an exception which should be passed to the caller.